## BUS 243 Mini Project 1: Analysis of Decline in User Engagement at eLinks

The objective of this study is to examine and quantify the recent drop in engagement of users on eLinks. Additionally, this study aims to develop and test different hypotheses for a decline in engagement and make recommendations for addressing the decline in engagement.

In [0]:
from pyspark.sql.types import *

### Data Preparation

#### Loading Users Data and Creating a Temporary Table

In [0]:
# File location and type
file_location = "/FileStore/tables/users.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df_users = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

In [0]:
# Create a user data table
temp_table_name = "users_csv"
df_users.createOrReplaceTempView(temp_table_name)

#### Loading Emails Data and Creating a Temporary Table

In [0]:
# File location and type
file_location = "/FileStore/tables/emails.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df_emails = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

In [0]:
# Create an emails data table
temp_table_name = "emails_csv"
df_emails.createOrReplaceTempView(temp_table_name)

#### Loading Events Data and Creating a Temporary Table

In [0]:
# File location and type
file_location = "/FileStore/tables/events.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df_events = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

In [0]:
# Create an events table
temp_table_name = "events_csv"
df_events.createOrReplaceTempView(temp_table_name)

### Viewing the Data

#### Users Data

In [0]:
%sql 
select 
cast(user_id as int),
cast(created_at as timestamp), 
cast(company_id as int), 
language, 
cast(activated_at as timestamp),
state
from users_csv limit 5

user_id created_at company_id language activated_at state 0 2013-01-01T20:59:39.000+0000 5737 english 2013-01-01T21:01:07.000+0000 active 1 2013-01-01T13:07:46.000+0000 28 english null pending 2 2013-01-01T10:59:05.000+0000 51 english null pending 3 2013-01-01T18:40:36.000+0000 2800 german 2013-01-01T18:42:02.000+0000 active 4 2013-01-01T14:37:51.000+0000 5110 indian 2013-01-01T14:39:05.000+0000 active

#### Emails Data

In [0]:
%sql 
select 
cast(user_id as int), 
cast(occurred_at as timestamp), 
action, 
cast(user_type as int)
from emails_csv limit 5

user_id occurred_at action user_type 0 2014-05-06T09:30:00.000+0000 sent_weekly_digest 1 0 2014-05-13T09:30:00.000+0000 sent_weekly_digest 1 0 2014-05-20T09:30:00.000+0000 sent_weekly_digest 1 0 2014-05-27T09:30:00.000+0000 sent_weekly_digest 1 0 2014-06-03T09:30:00.000+0000 sent_weekly_digest 1

#### Events Data

In [0]:
%sql 
select 
cast(user_id as int), 
cast(occurred_at as timestamp), 
event_type, 
event_name, 
location, 
device, 
cast(user_type as int)
from events_csv limit 5

user_id occurred_at event_type event_name location device user_type 10522 2014-05-02T11:02:39.000+0000 engagement login Japan dell inspiron notebook 3 10522 2014-05-02T11:02:53.000+0000 engagement home_page Japan dell inspiron notebook 3 10522 2014-05-02T11:03:28.000+0000 engagement like_message Japan dell inspiron notebook 3 10522 2014-05-02T11:04:09.000+0000 engagement view_inbox Japan dell inspiron notebook 3 10522 2014-05-02T11:03:16.000+0000 engagement search_run Japan dell inspiron notebook 3

### Data Exploration

#### Total Number of Engagements Per Day

The graph below shows the total number of engagement activities per day for the entire dataset. As shown by the graph, there is a moderate decline in the number of daily engagements starting at the end of July. To get a better idea of the trend of the data, we can also examine the total number of weekly engagements.

In [0]:
%sql 
SELECT 
DATE_TRUNC('day',occurred_at) AS day, 
COUNT(user_id) AS total_number_engagements
FROM events_csv 
GROUP BY day, event_type
HAVING event_type='engagement'

day total_number_engagements 2014-07-25T00:00:00.000+0000 1516 2014-06-07T00:00:00.000+0000 519 2014-05-07T00:00:00.000+0000 1361 2014-07-06T00:00:00.000+0000 216 2014-07-04T00:00:00.000+0000 1657 2014-07-17T00:00:00.000+0000 1265 2014-06-21T00:00:00.000+0000 637 2014-07-30T00:00:00.000+0000 1498 2014-05-26T00:00:00.000+0000 1129 2014-08-23T00:00:00.000+0000 417 2014-07-27T00:00:00.000+0000 410 2014-06-18T00:00:00.000+0000 1534 2014-08-11T00:00:00.000+0000 648 2014-05-16T00:00:00.000+0000 1660 2014-08-13T00:00:00.000+0000 1004 2014-08-27T00:00:00.000+0000 1011 2014-05-22T00:00:00.000+0000 1484 2014-07-12T00:00:00.000+0000 574 2014-08-04T00:00:00.000+0000 829 2014-08-22T00:00:00.000+0000 1143 2014-05-15T00:00:00.000+0000 1572 2014-06-13T00:00:00.000+0000 1482 2014-08-07T00:00:00.000+0000 1324 2014-08-31T00:00:00.000+0000 245 2014-05-28T00:00:00.000+0000 1344 2014-08-28T00:00:00.000+0000 1215 2014-06-29T00:00:00.000+0000 382 2014-05-20T00:00:00.000+0000 1377 2014-05-21T00:00:00.000+0000 1373 2014-07-20T00:00:00.000+0000 232 2014-07-18T00:00:00.000+0000 1805 2014-07-31T00:00:00.000+0000 1543 2014-06-17T00:00:00.000+0000 1457 2014-05-13T00:00:00.000+0000 1395 2014-05-27T00:00:00.000+0000 1366 2014-07-28T00:00:00.000+0000 1322 2014-07-15T00:00:00.000+0000 1387 2014-06-05T00:00:00.000+0000 1402 2014-06-12T00:00:00.000+0000 1504 2014-05-04T00:00:00.000+0000 271 2014-06-06T00:00:00.000+0000 1676 2014-06-10T00:00:00.000+0000 1299 2014-06-03T00:00:00.000+0000 1287 2014-05-08T00:00:00.000+0000 1496 2014-07-02T00:00:00.000+0000 1373 2014-05-30T00:00:00.000+0000 1509 2014-06-09T00:00:00.000+0000 994 2014-06-27T00:00:00.000+0000 1569 2014-05-31T00:00:00.000+0000 449 2014-07-01T00:00:00.000+0000 1386 2014-06-15T00:00:00.000+0000 351 2014-07-09T00:00:00.000+0000 1404 2014-07-14T00:00:00.000+0000 1314 2014-08-17T00:00:00.000+0000 203 2014-05-05T00:00:00.000+0000 1200 2014-08-06T00:00:00.000+0000 1236 2014-05-01T00:00:00.000+0000 1503 2014-06-20T00:00:00.000+0000 1461 2014-08-21T00:00:00.000+0000 1151 2014-05-06T00:00:00.000+0000 1494 2014-05-12T00:00:00.000+0000 1025 2014-07-26T00:00:00.000+0000 418 2014-06-02T00:00:00.000+0000 1240 2014-05-19T00:00:00.000+0000 1220 2014-07-23T00:00:00.000+0000 1501 2014-05-17T00:00:00.000+0000 486 2014-08-18T00:00:00.000+0000 972 2014-06-11T00:00:00.000+0000 1356 2014-07-24T00:00:00.000+0000 1439 2014-06-28T00:00:00.000+0000 466 2014-08-16T00:00:00.000+0000 428 2014-08-03T00:00:00.000+0000 303 2014-06-23T00:00:00.000+0000 1225 2014-06-24T00:00:00.000+0000 1372 2014-08-02T00:00:00.000+0000 614 2014-05-29T00:00:00.000+0000 1499 2014-05-02T00:00:00.000+0000 1969 2014-08-15T00:00:00.000+0000 1309 2014-08-12T00:00:00.000+0000 1021 2014-07-07T00:00:00.000+0000 1274 2014-05-18T00:00:00.000+0000 221 2014-08-29T00:00:00.000+0000 1173 2014-05-25T00:00:00.000+0000 304 2014-05-23T00:00:00.000+0000 1515 2014-07-11T00:00:00.000+0000 1402 2014-07-05T00:00:00.000+0000 583 2014-05-14T00:00:00.000+0000 1620 2014-08-25T00:00:00.000+0000 914 2014-06-08T00:00:00.000+0000 295 2014-07-08T00:00:00.000+0000 1377 2014-05-11T00:00:00.000+0000 319 2014-08-01T00:00:00.000+0000 1540 2014-07-13T00:00:00.000+0000 355 2014-06-16T00:00:00.000+0000 1281 2014-07-22T00:00:00.000+0000 1195 2014-08-26T00:00:00.000+0000 1022 2014-07-10T00:00:00.000+0000 1393 2014-07-29T00:00:00.000+0000 1471 2014-08-08T00:00:00.000+0000 1268 2014-08-05T00:00:00.000+0000 1136 2014-05-24T00:00:00.000+0000 558 2014-06-30T00:00:00.000+0000 1222 2014-08-30T00:00:00.000+0000 458 2014-06-14T00:00:00.000+0000 532 2014-06-22T00:00:00.000+0000 183 2014-06-19T00:00:00.000+0000 1405 2014-08-09T00:00:00.000+0000 473 2014-07-21T00:00:00.000+0000 1112 2014-05-10T00:00:00.000+0000 449 2014-08-20T00:00:00.000+0000 1225 2014-05-03T00:00:00.000+0000 442 2014-08-24T00:00:00.000+0000 186 2014-08-14T00:00:00.000+0000 1309 2014-05-09T00:00:00.000+0000 1468 2014-06-01T00:00:00.000+0000 294 2014-07-03T00:00:00.000+0000 1566 2014-06-26T00:00:00.000+0000 1454 2014-07-16T00:00:00.000+0000 151

#### Total Number of Engagements Per Week

When we look at the total number of engagements per week, we can see a clear drop in engagements starting at the end of July (July 27th). For the week of July 27th, the number of weekly engagements was 8,291, and for the week of August 24th it was 6,038. This is a weekly decline in engagement activity of over 2000. This suggests that an event that led to a decline in engagement may have occurred at the end of July. We should also examine the user data to see if we see a similar decline in the number of registered users in the same time period.

In [0]:
%sql 
SELECT 
DATE_TRUNC('week',occurred_at) AS week, 
COUNT(user_id) AS total_number_engagements
FROM events_csv 
GROUP BY week, event_type
HAVING event_type='engagement'

week total_number_engagements 2014-05-26T00:00:00.000+0000 7590 2014-08-11T00:00:00.000+0000 5922 2014-08-04T00:00:00.000+0000 6425 2014-07-28T00:00:00.000+0000 8291 2014-06-09T00:00:00.000+0000 7518 2014-07-14T00:00:00.000+0000 8070 2014-05-05T00:00:00.000+0000 7787 2014-05-12T00:00:00.000+0000 7979 2014-06-02T00:00:00.000+0000 8133 2014-05-19T00:00:00.000+0000 7831 2014-08-18T00:00:00.000+0000 6004 2014-06-23T00:00:00.000+0000 7639 2014-07-07T00:00:00.000+0000 7779 2014-08-25T00:00:00.000+0000 6038 2014-06-16T00:00:00.000+0000 7958 2014-06-30T00:00:00.000+0000 8003 2014-07-21T00:00:00.000+0000 7591 2014-04-28T00:00:00.000+0000 4185

#### All Users and Activated Users

The chart below shows the total number of all users and activated users per week. The graph shows that there is a dip in both the number of activated and all users between July 27th and August 10th, however there is overall positive growth in the number of both activate and all users. We would expect the number of users to be higher than the number of activated users, as users may sign up for eLinks and later change their mind, or forget to activate their account. Interestingly, people signed up for eLinks starting in January 2013, however the first engagement activity didn't take place until May 1st 2014, over a year after users started signing up for eLinks. This could be due to the critical mass of users required for social media networks to operate effectively.

In [0]:
%sql
SELECT DATE_TRUNC('week',created_at) AS week, COUNT(*) AS all_users,
	COUNT(CASE WHEN activated_at IS NOT NULL THEN u.user_id ELSE NULL END) AS activated_users
FROM `users_csv` u GROUP BY 1 ORDER BY 1

week all_users activated_users 2012-12-31T00:00:00.000+0000 56 26 2013-01-07T00:00:00.000+0000 67 29 2013-01-14T00:00:00.000+0000 77 47 2013-01-21T00:00:00.000+0000 77 36 2013-01-28T00:00:00.000+0000 75 30 2013-02-04T00:00:00.000+0000 87 48 2013-02-11T00:00:00.000+0000 81 41 2013-02-18T00:00:00.000+0000 82 39 2013-02-25T00:00:00.000+0000 81 33 2013-03-04T00:00:00.000+0000 85 43 2013-03-11T00:00:00.000+0000 87 33 2013-03-18T00:00:00.000+0000 95 32 2013-03-25T00:00:00.000+0000 93 33 2013-04-01T00:00:00.000+0000 86 40 2013-04-08T00:00:00.000+0000 96 35 2013-04-15T00:00:00.000+0000 93 42 2013-04-22T00:00:00.000+0000 101 48 2013-04-29T00:00:00.000+0000 100 48 2013-05-06T00:00:00.000+0000 107 45 2013-05-13T00:00:00.000+0000 107 55 2013-05-20T00:00:00.000+0000 104 41 2013-05-27T00:00:00.000+0000 113 49 2013-06-03T00:00:00.000+0000 113 51 2013-06-10T00:00:00.000+0000 115 51 2013-06-17T00:00:00.000+0000 120 46 2013-06-24T00:00:00.000+0000 126 57 2013-07-01T00:00:00.000+0000 128 57 2013-07-08T00:00:00.000+0000 132 52 2013-07-15T00:00:00.000+0000 142 71 2013-07-22T00:00:00.000+0000 129 66 2013-07-29T00:00:00.000+0000 140 69 2013-08-05T00:00:00.000+0000 132 66 2013-08-12T00:00:00.000+0000 148 73 2013-08-19T00:00:00.000+0000 151 71 2013-08-26T00:00:00.000+0000 150 79 2013-09-02T00:00:00.000+0000 165 65 2013-09-09T00:00:00.000+0000 163 71 2013-09-16T00:00:00.000+0000 164 84 2013-09-23T00:00:00.000+0000 167 92 2013-09-30T00:00:00.000+0000 179 81 2013-10-07T00:00:00.000+0000 175 88 2013-10-14T00:00:00.000+0000 173 74 2013-10-21T00:00:00.000+0000 190 97 2013-10-28T00:00:00.000+0000 195 92 2013-11-04T00:00:00.000+0000 195 97 2013-11-11T00:00:00.000+0000 193 94 2013-11-18T00:00:00.000+0000 177 82 2013-11-25T00:00:00.000+0000 208 103 2013-12-02T00:00:00.000+0000 214 96 2013-12-09T00:00:00.000+0000 216 117 2013-12-16T00:00:00.000+0000 221 123 2013-12-23T00:00:00.000+0000 233 104 2013-12-30T00:00:00.000+0000 233 132 2014-01-06T00:00:00.000+0000 232 122 2014-01-13T00:00:00.000+0000 224 112 2014-01-20T00:00:00.000+0000 249 113 2014-01-27T00:00:00.000+0000 246 130 2014-02-03T00:00:00.000+0000 257 132 2014-02-10T00:00:00.000+0000 262 135 2014-02-17T00:00:00.000+0000 272 127 2014-02-24T00:00:00.000+0000 268 127 2014-03-03T00:00:00.000+0000 272 135 2014-03-10T00:00:00.000+0000 288 152 2014-03-17T00:00:00.000+0000 286 132 2014-03-24T00:00:00.000+0000 300 151 2014-03-31T00:00:00.000+0000 306 161 2014-04-07T00:00:00.000+0000 319 166 2014-04-14T00:00:00.000+0000 303 165 2014-04-21T00:00:00.000+0000 365 176 2014-04-28T00:00:00.000+0000 353 172 2014-05-05T00:00:00.000+0000 350 160 2014-05-12T00:00:00.000+0000 362 186 2014-05-19T00:00:00.000+0000 371 177 2014-05-26T00:00:00.000+0000 366 186 2014-06-02T00:00:00.000+0000 390 197 2014-06-09T00:00:00.000+0000 413 198 2014-06-16T00:00:00.000+0000 421 222 2014-06-23T00:00:00.000+0000 404 210 2014-06-30T00:00:00.000+0000 405 199 2014-07-07T00:00:00.000+0000 424 223 2014-07-14T00:00:00.000+0000 426 215 2014-07-21T00:00:00.000+0000 458 228 2014-07-28T00:00:00.000+0000 476 234 2014-08-04T00:00:00.000+0000 406 189 2014-08-11T00:00:00.000+0000 473 250 2014-08-18T00:00:00.000+0000 468 259 2014-08-25T00:00:00.000+0000 514 266

As shown below, the first engagement activity took place on May 1st 2014, and the last engagement activity took place at the end of August.

In [0]:
%sql 
SELECT 
MIN(occurred_at) AS first_engagement,
MAX(occurred_at) AS last_engagement
FROM events_csv
WHERE event_type='engagement'

first_engagement last_engagement 2014-05-01 02:27:15 2014-08-31 20:23:22

### Hypotheses

The following Hypotheses will be tested: 
1. **Hypothesis 1:** Users Who Joined eLinks a Long Time Ago Are No Longer Engaging
2. **Hypothesis 2:** Certain Elements of the Website Were Non-Functional, Resulting in a Decline in Engagement
3. **Hypothesis 3:** Some Users Have Switched Over to a Different Platform
4. **Hypothesis 4:** eLinks Changed Their Engagement Techniques

#### Hypothesis 1: Users Who Joined eLinks a Long Time Ago Are No Longer Engaging

The first hypothesis to test is that older users are no longer engaging with the platform, which is why the engagement rate has declined recently. This hypothesis was developed based on the logic that some users who joined the platform a long time ago may have forgotten that they ever joined the platform, or after using the platform for some time decided to switch to a different platform if eLinks didn’t meet their expectations.

In [0]:
%sql 

SELECT 
day_of_activation,
last_engagement_for_user,
last_engagement,
CASE 
WHEN last_engagement_for_user IS NULL THEN DATEDIFF(last_engagement,day_of_activation) 
ELSE DATEDIFF(last_engagement,last_engagement_for_user) 
END AS days_since_last_engagement,
DATEDIFF(last_engagement,day_of_activation) AS days_on_platform

FROM(
SELECT 
CAST(u.user_id AS int), 
DATE_TRUNC('day',u.activated_at) AS day_of_activation, 
MAX(DATE_TRUNC('day',e.occurred_at)) AS last_engagement_for_user, 
(SELECT 
MAX(DATE_TRUNC('day',occurred_at)) AS last_engagement
FROM events_csv
WHERE event_type='engagement') AS last_engagement
FROM users_csv u LEFT JOIN events_csv e ON 
u.user_id = e.user_id
GROUP BY u.user_id, u.activated_at
HAVING u.activated_at IS NOT NULL
ORDER BY u.user_id ASC)

day_of_activation last_engagement_for_user last_engagement days_since_last_engagement days_on_platform 2013-01-01T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 607 607 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-24T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 160 160 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2013-01-10T00:00:00.000+0000 2014-08-30T00:00:00.000+0000 2014-08-31T00:00:00.000+0000 1 598 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-25T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 159 159 2014-03-26T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 158 158 2013-03-28T00:00:00.000+0000 2014-08-23T00:00:00.000+0000 2014-08-31T00:00:00.000+0000 8 521 2014-03-26T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 158 158 2014-03-26T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 158 158 2014-03-26T00:00:00.000+0000 null 2014-08-31T00:00:00.000+0000 158 158 2014-03-26T00:00:00

As shown by the scatter plot above, there does not appear to be a relationship between the number of days a user has been on the platform and how recently they last engaged on eLinks. Based on the data shown above, there appear to be three main clusters of users:


  1. **New users with moderate activity:**
        Users who are new to the platform (less than 150 days) and engaged 100-150 days ago 
  2. **Long-term users with high-moderate activity:**
        Users who have been on the platform a long time (500-550 days) and have engaged within the last 100 days 
  3. **Long-term users who have been inactive for a long time:**
        Users who have been on the platform a long time (500-600 days) and last engaged 500-600 days ago

#### Hypothesis 2: Certain Elements of the Website Were Non-Functional, Resulting in a Decline in Engagement

The second hypothesis is that certain elements of the website were non-functional, resulting in a decline in engagement. For example, if users were not able to “like” posts by other users due to a bug on the website, this would result in a decline in engagement activities. While this cannot be tested directly, we can examine the number of different engagement activities and look for a sharp decline in one or more of them.

In [0]:
%sql 
SELECT 
DATE_TRUNC('week',occurred_at) AS week, 
COUNT(user_id) AS total_number_engagements,
event_name
FROM events_csv 
GROUP BY week, event_type, event_name
HAVING event_type='engagement'

week total_number_engagements event_name 2014-07-14T00:00:00.000+0000 16 search_click_result_10 2014-07-14T00:00:00.000+0000 24 search_click_result_3 2014-07-14T00:00:00.000+0000 15 search_click_result_7 2014-08-25T00:00:00.000+0000 382 search_autocomplete 2014-06-09T00:00:00.000+0000 891 login 2014-06-09T00:00:00.000+0000 1322 view_inbox 2014-06-16T00:00:00.000+0000 43 search_click_result_2 2014-07-28T00:00:00.000+0000 31 search_click_result_3 2014-08-25T00:00:00.000+0000 1108 view_inbox 2014-06-02T00:00:00.000+0000 2412 home_page 2014-06-16T00:00:00.000+0000 37 search_click_result_1 2014-07-21T00:00:00.000+0000 27 search_click_result_3 2014-08-11T00:00:00.000+0000 274 search_run 2014-05-26T00:00:00.000+0000 22 search_click_result_9 2014-07-21T00:00:00.000+0000 1335 view_inbox 2014-06-16T00:00:00.000+0000 14 search_click_result_9 2014-08-11T00:00:00.000+0000 22 search_click_result_10 2014-07-14T00:00:00.000+0000 18 search_click_result_6 2014-08-04T00:00:00.000+0000 10 search_click_result_8 2014-05-12T00:00:00.000+0000 32 search_click_result_1 2014-04-28T00:00:00.000+0000 12 search_click_result_8 2014-06-09T00:00:00.000+0000 294 search_run 2014-07-07T00:00:00.000+0000 901 login 2014-08-11T00:00:00.000+0000 1069 like_message 2014-07-21T00:00:00.000+0000 2229 home_page 2014-08-18T00:00:00.000+0000 10 search_click_result_7 2014-06-02T00:00:00.000+0000 36 search_click_result_2 2014-06-16T00:00:00.000+0000 11 search_click_result_10 2014-08-04T00:00:00.000+0000 16 search_click_result_9 2014-05-19T00:00:00.000+0000 856 send_message 2014-05-19T00:00:00.000+0000 1297 view_inbox 2014-08-18T00:00:00.000+0000 11 search_click_result_6 2014-07-28T00:00:00.000+0000 862 send_message 2014-08-25T00:00:00.000+0000 11 search_click_result_2 2014-05-12T00:00:00.000+0000 18 search_click_result_7 2014-07-28T00:00:00.000+0000 1544 like_message 2014-05-19T00:00:00.000+0000 2273 home_page 2014-07-21T00:00:00.000+0000 24 search_click_result_9 2014-07-07T00:00:00.000+0000 1426 view_inbox 2014-05-19T00:00:00.000+0000 390 search_autocomplete 2014-06-16T00:00:00.000+0000 871 send_message 2014-06-30T00:00:00.000+0000 2320 home_page 2014-05-12T00:00:00.000+0000 849 send_message 2014-05-19T00:00:00.000+0000 52 search_click_result_2 2014-06-16T00:00:00.000+0000 9 search_click_result_6 2014-08-25T00:00:00.000+0000 7 search_click_result_9 2014-08-11T00:00:00.000+0000 21 search_click_result_6 2014-06-30T00:00:00.000+0000 822 send_message 2014-07-14T00:00:00.000+0000 32 search_click_result_4 2014-07-14T00:00:00.000+0000 935 login 2014-08-04T00:00:00.000+0000 1205 like_message 2014-06-30T00:00:00.000+0000 28 search_click_result_5 2014-08-18T00:00:00.000+0000 13 search_click_result_5 2014-07-28T00:00:00.000+0000 17 search_click_result_6 2014-05-19T00:00:00.000+0000 22 search_click_result_6 2014-05-12T00:00:00.000+0000 924 login 2014-05-12T00:00:00.000+0000 31 search_click_result_4 2014-06-02T00:00:00.000+0000 11 search_click_result_10 2014-08-11T00:00:00.000+0000 16 search_click_result_8 2014-06-23T00:00:00.000+0000 14 search_click_result_6 2014-08-04T00:00:00.000+0000 203 search_run 2014-05-05T00:00:00.000+0000 35 search_click_result_2 2014-07-21T00:00:00.000+0000 910 login 2014-05-05T00:00:00.000+0000 830 send_message 2014-06-23T00:00:00.000+0000 26 search_click_result_4 2014-07-28T00:00:00.000+0000 21 search_click_result_10 2014-05-05T00:00:00.000+0000 320 search_run 2014-08-11T00:00:00.000+0000 735 login 2014-08-18T00:00:00.000+0000 590 send_message 2014-07-21T00:00:00.000+0000 38 search_click_result_1 2014-05-05T00:00:00.000+0000 25 search_click_result_3 2014-07-21T00:00:00.000+0000 37 search_click_result_2 2014-08-04T00:00:00.000+0000 9 search_click_result_7 2014-05-26T00:00:00.000+0000 1426 like_message 2014-05-26T00:00:00.000+0000 21 search_click_result_7 2014-06-23T00:00:00.000+0000 7 search_click_result_10 2014-05-05T00:00:00.000+0000 10 search_click_result_10 2014-08-04T00:00:00.000+0000 1900 home_page 2014-07-14T00:00:00.000+0000 43 search_click_result_2 2

As shown by the line graph above, the engagement activity with the highest decline is users loading the home page of eLinks. This does not support the hypothesis that certain features of the website were not operational, as it shows a decline in the number of people loading the website itself. There may have be an external factor that caused a decline in the number of people vising eLinks.

To make sure that the website was fully functional at the end of August, we can also look at the number of signup activities that took place in that time period. The graph below shows the number of signup activities per day. As shown, the number of sign up activities follows a cyclical pattern, with certain times of the week being more popular to sign up for eLinks. The graph shows usual sign up activity during the end of August, so we can conclude that the website was functioning normally.

In [0]:
%sql
SELECT 
CAST(DATE_TRUNC('day',occurred_at) AS timestamp) AS day,
COUNT(user_id) AS number_signup_activities
FROM events_csv
WHERE event_type ='signup_flow'
GROUP BY day
ORDER BY day ASC

day number_signup_activities 2014-05-01T00:00:00.000+0000 187 2014-05-02T00:00:00.000+0000 151 2014-05-03T00:00:00.000+0000 47 2014-05-04T00:00:00.000+0000 55 2014-05-05T00:00:00.000+0000 140 2014-05-06T00:00:00.000+0000 147 2014-05-07T00:00:00.000+0000 172 2014-05-08T00:00:00.000+0000 176 2014-05-09T00:00:00.000+0000 165 2014-05-10T00:00:00.000+0000 46 2014-05-11T00:00:00.000+0000 38 2014-05-12T00:00:00.000+0000 150 2014-05-13T00:00:00.000+0000 190 2014-05-14T00:00:00.000+0000 171 2014-05-15T00:00:00.000+0000 185 2014-05-16T00:00:00.000+0000 177 2014-05-17T00:00:00.000+0000 43 2014-05-18T00:00:00.000+0000 44 2014-05-19T00:00:00.000+0000 165 2014-05-20T00:00:00.000+0000 206 2014-05-21T00:00:00.000+0000 133 2014-05-22T00:00:00.000+0000 176 2014-05-23T00:00:00.000+0000 183 2014-05-24T00:00:00.000+0000 48 2014-05-25T00:00:00.000+0000 44 2014-05-26T00:00:00.000+0000 144 2014-05-27T00:00:00.000+0000 166 2014-05-28T00:00:00.000+0000 190 2014-05-29T00:00:00.000+0000 165 2014-05-30T00:00:00.000+0000 202 2014-05-31T00:00:00.000+0000 50 2014-06-01T00:00:00.000+0000 61 2014-06-02T00:00:00.000+0000 179 2014-06-03T00:00:00.000+0000 161 2014-06-04T00:00:00.000+0000 213 2014-06-05T00:00:00.000+0000 184 2014-06-06T00:00:00.000+0000 201 2014-06-07T00:00:00.000+0000 43 2014-06-08T00:00:00.000+0000 62 2014-06-09T00:00:00.000+0000 185 2014-06-10T00:00:00.000+0000 175 2014-06-11T00:00:00.000+0000 200 2014-06-12T00:00:00.000+0000 204 2014-06-13T00:00:00.000+0000 195 2014-06-14T00:00:00.000+0000 44 2014-06-15T00:00:00.000+0000 70 2014-06-16T00:00:00.000+0000 210 2014-06-17T00:00:00.000+0000 230 2014-06-18T00:00:00.000+0000 225 2014-06-19T00:00:00.000+0000 171 2014-06-20T00:00:00.000+0000 203 2014-06-21T00:00:00.000+0000 49 2014-06-22T00:00:00.000+0000 48 2014-06-23T00:00:00.000+0000 219 2014-06-24T00:00:00.000+0000 130 2014-06-25T00:00:00.000+0000 191 2014-06-26T00:00:00.000+0000 206 2014-06-27T00:00:00.000+0000 219 2014-06-28T00:00:00.000+0000 62 2014-06-29T00:00:00.000+0000 54 2014-06-30T00:00:00.000+0000 199 2014-07-01T00:00:00.000+0000 207 2014-07-02T00:00:00.000+0000 201 2014-07-03T00:00:00.000+0000 171 2014-07-04T00:00:00.000+0000 184 2014-07-05T00:00:00.000+0000 49 2014-07-06T00:00:00.000+0000 46 2014-07-07T00:00:00.000+0000 240 2014-07-08T00:00:00.000+0000 192 2014-07-09T00:00:00.000+0000 197 2014-07-10T00:00:00.000+0000 183 2014-07-11T00:00:00.000+0000 232 2014-07-12T00:00:00.000+0000 50 2014-07-13T00:00:00.000+0000 53 2014-07-14T00:00:00.000+0000 206 2014-07-15T00:00:00.000+0000 202 2014-07-16T00:00:00.000+0000 267 2014-07-17T00:00:00.000+0000 154 2014-07-18T00:00:00.000+0000 198 2014-07-19T00:00:00.000+0000 52 2014-07-20T00:00:00.000+0000 51 2014-07-21T00:00:00.000+0000 196 2014-07-22T00:00:00.000+0000 204 2014-07-23T00:00:00.000+0000 235 2014-07-24T00:00:00.000+0000 207 2014-07-25T00:00:00.000+0000 212 2014-07-26T00:00:00.000+0000 61 2014-07-27T00:00:00.000+0000 77 2014-07-28T00:00:00.000+0000 225 2014-07-29T00:00:00.000+0000 187 2014-07-30T00:00:00.000+0000 233 2014-07-31T00:00:00.000+0000 240 2014-08-01T00:00:00.000+0000 213 2014-08-02T00:00:00.000+0000 67 2014-08-03T00:00:00.000+0000 63 2014-08-04T00:00:00.000+0000 144 2014-08-05T00:00:00.000+0000 197 2014-08-06T00:00:00.000+0000 170 2014-08-07T00:00:00.000+0000 206 2014-08-08T00:00:00.000+0000 185 2014-08-09T00:00:00.000+0000 64 2014-08-10T00:00:00.000+0000 51 2014-08-11T00:00:00.000+0000 231 2014-08-12T00:00:00.000+0000 185 2014-08-13T00:00:00.000+0000 253 2014-08-14T00:00:00.000+0000 237 2014-08-15T00:00:00.000+0000 231 2014-08-16T00:00:00.000+0000 72 2014-08-17T00:00:00.000+0000 61 2014-08-18T00:00:00.000+0000 243 2014-08-19T00:00:00.000+0000 220 2014-08-20T00:00:00.000+0000 219 2014-08-21T00:00:00.000+0000 234 2014-08-22T00:00:00.000+0000 255 2014-08-23T00:00:00.000+0000 68 2014-08-24T00:00:00.000+0000 51 2014-08-25T00:00:00.000+0000 254 2014-08-26T00:00:00.000+0000 211 2014-08-27T00:00:00.000+0000 249 2014-08-28T00:00:00.000+0000 256 2014-08-29T00:00:00.000+0000 251 2014-08-30T0

#### Hypothesis 3: Some Users Have Switched Over to a Different Platform

The third hypothesis is that there is a decline in engagement due to emergence of a competitor networking site. We can look for evidence for this hypothesis by examining whether there is a substantial decline in activity from users from a certain region or company, as new websites typically become popular regionally before spreading to other areas, and some companies encourage their employees to use a specific networking site. If your company recommends that you use a certain site and all your colleagues are using that site, it is likely that you would switch over to that site.

In [0]:
%sql

SELECT 
CAST(user_id AS int),
DATE_TRUNC('week',occurred_at) AS week, 
COUNT(user_id) AS total_number_engagements, 
location
FROM events_csv  
GROUP BY week, event_type, user_id, location
HAVING event_type='engagement'

user_id week total_number_engagements location 12350 2014-05-12T00:00:00.000+0000 3 Australia 12402 2014-05-12T00:00:00.000+0000 2 Italy 12492 2014-05-12T00:00:00.000+0000 9 United States 12589 2014-06-02T00:00:00.000+0000 3 Canada 12689 2014-05-19T00:00:00.000+0000 2 Sweden 13046 2014-05-26T00:00:00.000+0000 3 Japan 13267 2014-05-26T00:00:00.000+0000 6 United Kingdom 13765 2014-06-02T00:00:00.000+0000 3 United States 13979 2014-06-09T00:00:00.000+0000 15 Poland 14277 2014-06-16T00:00:00.000+0000 9 Russia 14319 2014-06-16T00:00:00.000+0000 9 United Kingdom 14643 2014-06-23T00:00:00.000+0000 6 Indonesia 15587 2014-07-07T00:00:00.000+0000 13 South Africa 16220 2014-07-14T00:00:00.000+0000 7 Indonesia 16222 2014-07-14T00:00:00.000+0000 13 United States 16635 2014-07-21T00:00:00.000+0000 5 Brazil 16761 2014-07-28T00:00:00.000+0000 12 Russia 16872 2014-07-28T00:00:00.000+0000 9 United States 17027 2014-07-28T00:00:00.000+0000 12 Germany 17703 2014-08-11T00:00:00.000+0000 4 Japan 17806 2014-08-25T00:00:00.000+0000 3 United States 18076 2014-08-11T00:00:00.000+0000 10 Spain 18473 2014-08-18T00:00:00.000+0000 8 Belgium 12473 2014-05-26T00:00:00.000+0000 7 United States 12488 2014-05-12T00:00:00.000+0000 2 United States 12611 2014-05-12T00:00:00.000+0000 3 United States 13179 2014-05-26T00:00:00.000+0000 2 United States 13325 2014-05-26T00:00:00.000+0000 17 United States 13453 2014-06-09T00:00:00.000+0000 6 Japan 14282 2014-06-16T00:00:00.000+0000 5 India 14362 2014-06-16T00:00:00.000+0000 15 United States 14478 2014-06-16T00:00:00.000+0000 4 United States 14658 2014-06-23T00:00:00.000+0000 6 United States 14691 2014-06-23T00:00:00.000+0000 2 Nigeria 15168 2014-07-14T00:00:00.000+0000 2 United States 15498 2014-07-07T00:00:00.000+0000 7 India 15903 2014-07-14T00:00:00.000+0000 31 United States 16388 2014-07-21T00:00:00.000+0000 6 United States 17490 2014-08-04T00:00:00.000+0000 8 United States 18104 2014-08-18T00:00:00.000+0000 9 India 12270 2014-05-05T00:00:00.000+0000 4 United Arab Emirates 12389 2014-05-12T00:00:00.000+0000 14 Brazil 12545 2014-05-12T00:00:00.000+0000 4 United States 12833 2014-05-19T00:00:00.000+0000 13 United States 12887 2014-05-19T00:00:00.000+0000 2 United States 14011 2014-06-09T00:00:00.000+0000 7 United States 14442 2014-06-16T00:00:00.000+0000 3 Italy 14502 2014-06-23T00:00:00.000+0000 14 United States 14670 2014-06-23T00:00:00.000+0000 2 Indonesia 15128 2014-06-30T00:00:00.000+0000 2 Mexico 15531 2014-07-07T00:00:00.000+0000 5 Spain 16787 2014-07-28T00:00:00.000+0000 10 United States 16992 2014-07-28T00:00:00.000+0000 5 Switzerland 17231 2014-08-04T00:00:00.000+0000 4 Japan 17817 2014-08-11T00:00:00.000+0000 9 United States 18439 2014-08-18T00:00:00.000+0000 9 Japan 11849 2014-04-28T00:00:00.000+0000 13 United States 12029 2014-05-05T00:00:00.000+0000 18 United Arab Emirates 12844 2014-05-19T00:00:00.000+0000 7 Switzerland 13044 2014-05-26T00:00:00.000+0000 14 United Kingdom 13098 2014-05-26T00:00:00.000+0000 6 United States 13252 2014-06-23T00:00:00.000+0000 9 Mexico 13706 2014-06-23T00:00:00.000+0000 3 Japan 15233 2014-06-30T00:00:00.000+0000 3 United States 15253 2014-06-30T00:00:00.000+0000 18 United Kingdom 15517 2014-07-07T00:00:00.000+0000 7 Iran 16182 2014-07-14T00:00:00.000+0000 3 United States 16289 2014-07-21T00:00:00.000+0000 4 Australia 16435 2014-07-21T00:00:00.000+0000 4 Australia 17081 2014-08-04T00:00:00.000+0000 10 United States 17298 2014-08-04T00:00:00.000+0000 11 United States 17545 2014-08-04T00:00:00.000+0000 8 Greece 17608 2014-08-04T00:00:00.000+0000 2 Japan 18844 2014-08-25T00:00:00.000+0000 2 Hong Kong 19055 2014-08-25T00:00:00.000+0000 2 Germany 30 2014-08-11T00:00:00.000+0000 5 Australia 101 2014-06-02T00:00:00.000+0000 8 Netherlands 11778 2014-05-26T00:00:00.000+0000 4 Indonesia 12296 2014-05-12T00:00:00.000+0000 9 South Africa 12315 2014-05-12T00:00:00.000+0000 6 Indonesia 13617 2014-06-02T00:00:00.000+0000 6 United States 14012 2014-06-09T00:00:00.000+0000 5 United States 14

As shown by the line graph above, Germany is the country with the largest number of engagements. There is a significant drop in the number of weekly engagements from 694 in the week of July 20th, 2014 to 355 in the week of August 24th, 2014. Other than May, the July value is the lowest that the weekly number of engagements has ever been in Germany. We don't consider the May value to be the low point because this is the start of engagement activity across the dataset.

France is the country with the second highest number of engagements. As shown by the graph, there is gradual decline from 496 weekly engagements in the week of June 29th to 274 weekly engagements in the week of August 24th. As Germany and France are geographically close, and both have a significant decline in engagement activity, it's possible that a competitor networking website has emerged in this region.

Further analysis is required to determine the cause for the significant drop in engagement of users located in Germany and France. Next, we can look at the engagement activitiy for all countries except Germany and France, to see if the decline in the engagement in these countries accounts for the decline in the overall weekly engagement.

In [0]:
%sql
SELECT 
CAST(user_id AS int),
DATE_TRUNC('week',occurred_at) AS week, 
COUNT(user_id) AS total_number_engagements, 
location
FROM events_csv
WHERE location NOT IN ('Germany','France')
GROUP BY week, event_type, user_id, location
HAVING event_type='engagement'

user_id week total_number_engagements location 12350 2014-05-12T00:00:00.000+0000 3 Australia 12402 2014-05-12T00:00:00.000+0000 2 Italy 12492 2014-05-12T00:00:00.000+0000 9 United States 12589 2014-06-02T00:00:00.000+0000 3 Canada 12689 2014-05-19T00:00:00.000+0000 2 Sweden 13046 2014-05-26T00:00:00.000+0000 3 Japan 13267 2014-05-26T00:00:00.000+0000 6 United Kingdom 13765 2014-06-02T00:00:00.000+0000 3 United States 13979 2014-06-09T00:00:00.000+0000 15 Poland 14277 2014-06-16T00:00:00.000+0000 9 Russia 14319 2014-06-16T00:00:00.000+0000 9 United Kingdom 14643 2014-06-23T00:00:00.000+0000 6 Indonesia 15587 2014-07-07T00:00:00.000+0000 13 South Africa 16220 2014-07-14T00:00:00.000+0000 7 Indonesia 16222 2014-07-14T00:00:00.000+0000 13 United States 16635 2014-07-21T00:00:00.000+0000 5 Brazil 16761 2014-07-28T00:00:00.000+0000 12 Russia 16872 2014-07-28T00:00:00.000+0000 9 United States 17703 2014-08-11T00:00:00.000+0000 4 Japan 17806 2014-08-25T00:00:00.000+0000 3 United States 18076 2014-08-11T00:00:00.000+0000 10 Spain 18473 2014-08-18T00:00:00.000+0000 8 Belgium 12473 2014-05-26T00:00:00.000+0000 7 United States 12488 2014-05-12T00:00:00.000+0000 2 United States 12611 2014-05-12T00:00:00.000+0000 3 United States 13179 2014-05-26T00:00:00.000+0000 2 United States 13325 2014-05-26T00:00:00.000+0000 17 United States 13453 2014-06-09T00:00:00.000+0000 6 Japan 14282 2014-06-16T00:00:00.000+0000 5 India 14362 2014-06-16T00:00:00.000+0000 15 United States 14478 2014-06-16T00:00:00.000+0000 4 United States 14658 2014-06-23T00:00:00.000+0000 6 United States 14691 2014-06-23T00:00:00.000+0000 2 Nigeria 15168 2014-07-14T00:00:00.000+0000 2 United States 15498 2014-07-07T00:00:00.000+0000 7 India 15903 2014-07-14T00:00:00.000+0000 31 United States 16388 2014-07-21T00:00:00.000+0000 6 United States 17490 2014-08-04T00:00:00.000+0000 8 United States 18104 2014-08-18T00:00:00.000+0000 9 India 12270 2014-05-05T00:00:00.000+0000 4 United Arab Emirates 12389 2014-05-12T00:00:00.000+0000 14 Brazil 12545 2014-05-12T00:00:00.000+0000 4 United States 12833 2014-05-19T00:00:00.000+0000 13 United States 12887 2014-05-19T00:00:00.000+0000 2 United States 14011 2014-06-09T00:00:00.000+0000 7 United States 14442 2014-06-16T00:00:00.000+0000 3 Italy 14502 2014-06-23T00:00:00.000+0000 14 United States 14670 2014-06-23T00:00:00.000+0000 2 Indonesia 15128 2014-06-30T00:00:00.000+0000 2 Mexico 15531 2014-07-07T00:00:00.000+0000 5 Spain 16787 2014-07-28T00:00:00.000+0000 10 United States 16992 2014-07-28T00:00:00.000+0000 5 Switzerland 17231 2014-08-04T00:00:00.000+0000 4 Japan 17817 2014-08-11T00:00:00.000+0000 9 United States 18439 2014-08-18T00:00:00.000+0000 9 Japan 11849 2014-04-28T00:00:00.000+0000 13 United States 12029 2014-05-05T00:00:00.000+0000 18 United Arab Emirates 12844 2014-05-19T00:00:00.000+0000 7 Switzerland 13044 2014-05-26T00:00:00.000+0000 14 United Kingdom 13098 2014-05-26T00:00:00.000+0000 6 United States 13252 2014-06-23T00:00:00.000+0000 9 Mexico 13706 2014-06-23T00:00:00.000+0000 3 Japan 15233 2014-06-30T00:00:00.000+0000 3 United States 15253 2014-06-30T00:00:00.000+0000 18 United Kingdom 15517 2014-07-07T00:00:00.000+0000 7 Iran 16182 2014-07-14T00:00:00.000+0000 3 United States 16289 2014-07-21T00:00:00.000+0000 4 Australia 16435 2014-07-21T00:00:00.000+0000 4 Australia 17081 2014-08-04T00:00:00.000+0000 10 United States 17298 2014-08-04T00:00:00.000+0000 11 United States 17545 2014-08-04T00:00:00.000+0000 8 Greece 17608 2014-08-04T00:00:00.000+0000 2 Japan 18844 2014-08-25T00:00:00.000+0000 2 Hong Kong 30 2014-08-11T00:00:00.000+0000 5 Australia 101 2014-06-02T00:00:00.000+0000 8 Netherlands 11778 2014-05-26T00:00:00.000+0000 4 Indonesia 12296 2014-05-12T00:00:00.000+0000 9 South Africa 12315 2014-05-12T00:00:00.000+0000 6 Indonesia 13617 2014-06-02T00:00:00.000+0000 6 United States 14012 2014-06-09T00:00:00.000+0000 5 United States 14187 2014-06-09T00:00:00.000+0000 25 Singapore 15113 2014-06-30T00:00:00.000+0000 5 Mexico 1

As shown by the line graph above, if we do not include the weekly engagement data from Germany and France, there is still a significant decline in the number of weekly engagements from the week of July 27th to August 24th. Therefore, the decline in engagement in Germany and France alone cannot account for the significant decrease in engagement.

Next, we can look at whether there is a sharp decline in engagement from users from specific companies to see if there may be a large company that changed their social media strategy and started encouraging their employees to use a different networking platform. If the company is large enough, such a change in strategy could negatively impact overall engagement on eLinks.

In [0]:
%sql 
SELECT 
week, 
COUNT(user_id) AS weekly_engagements, 
CAST(company_id AS int)

FROM

(SELECT
e.user_id, 
DATE_TRUNC('week',e.occurred_at) AS week, 
u.company_id
FROM events_csv e LEFT JOIN users_csv u
  ON e.user_id = u.user_id
WHERE e.event_type='engagement')

GROUP BY week, company_id

week weekly_engagements company_id 2014-05-05T00:00:00.000+0000 10 10544 2014-05-05T00:00:00.000+0000 5 68 2014-06-02T00:00:00.000+0000 5 7130 2014-06-09T00:00:00.000+0000 5 10 2014-06-09T00:00:00.000+0000 5 7467 2014-06-16T00:00:00.000+0000 62 4 2014-06-23T00:00:00.000+0000 4 7407 2014-07-14T00:00:00.000+0000 5 3194 2014-07-21T00:00:00.000+0000 23 13173 2014-07-28T00:00:00.000+0000 20 5650 2014-07-28T00:00:00.000+0000 8 7793 2014-07-28T00:00:00.000+0000 6 96 2014-08-04T00:00:00.000+0000 6 740 2014-08-18T00:00:00.000+0000 7 372 2014-08-04T00:00:00.000+0000 8 11395 2014-08-04T00:00:00.000+0000 2 527 2014-08-18T00:00:00.000+0000 13 4290 2014-08-18T00:00:00.000+0000 8 6428 2014-06-16T00:00:00.000+0000 9 4148 2014-06-16T00:00:00.000+0000 18 2296 2014-05-12T00:00:00.000+0000 40 925 2014-05-12T00:00:00.000+0000 8 9866 2014-05-19T00:00:00.000+0000 3 265 2014-05-26T00:00:00.000+0000 10 6232 2014-06-02T00:00:00.000+0000 12 12066 2014-06-09T00:00:00.000+0000 7 12472 2014-06-16T00:00:00.000+0000 9 69 2014-06-23T00:00:00.000+0000 2 11993 2014-06-23T00:00:00.000+0000 4 309 2014-06-30T00:00:00.000+0000 6 67 2014-07-14T00:00:00.000+0000 6 8825 2014-07-14T00:00:00.000+0000 30 9764 2014-07-28T00:00:00.000+0000 53 44 2014-07-28T00:00:00.000+0000 2 4240 2014-08-11T00:00:00.000+0000 6 8011 2014-08-18T00:00:00.000+0000 2 509 2014-08-18T00:00:00.000+0000 2 1115 2014-08-18T00:00:00.000+0000 8 8283 2014-04-28T00:00:00.000+0000 5 10595 2014-05-12T00:00:00.000+0000 3 3615 2014-06-09T00:00:00.000+0000 3 1715 2014-06-09T00:00:00.000+0000 8 2668 2014-06-23T00:00:00.000+0000 2 11442 2014-06-30T00:00:00.000+0000 2 4377 2014-07-14T00:00:00.000+0000 7 2817 2014-07-21T00:00:00.000+0000 3 564 2014-08-04T00:00:00.000+0000 8 6965 2014-08-11T00:00:00.000+0000 23 47 2014-08-18T00:00:00.000+0000 8 5430 2014-08-25T00:00:00.000+0000 6 13067 2014-08-25T00:00:00.000+0000 2 11092 2014-04-28T00:00:00.000+0000 155 3 2014-06-09T00:00:00.000+0000 18 18 2014-05-19T00:00:00.000+0000 16 10549 2014-06-02T00:00:00.000+0000 24 12786 2014-06-02T00:00:00.000+0000 3 2069 2014-06-02T00:00:00.000+0000 8 8037 2014-06-02T00:00:00.000+0000 10 9625 2014-06-23T00:00:00.000+0000 51 9 2014-06-23T00:00:00.000+0000 7 11030 2014-06-30T00:00:00.000+0000 2 291 2014-06-30T00:00:00.000+0000 3 4828 2014-07-14T00:00:00.000+0000 15 2044 2014-07-28T00:00:00.000+0000 10 5352 2014-07-28T00:00:00.000+0000 2 6949 2014-08-04T00:00:00.000+0000 4 12095 2014-08-25T00:00:00.000+0000 19 39 2014-08-18T00:00:00.000+0000 2 75 2014-08-18T00:00:00.000+0000 3 124 2014-04-28T00:00:00.000+0000 12 148 2014-05-05T00:00:00.000+0000 7 11385 2014-05-26T00:00:00.000+0000 8 7864 2014-06-09T00:00:00.000+0000 19 8258 2014-06-16T00:00:00.000+0000 4 1470 2014-08-11T00:00:00.000+0000 6 5045 2014-06-23T00:00:00.000+0000 4 10962 2014-07-07T00:00:00.000+0000 5 6123 2014-07-07T00:00:00.000+0000 13 9292 2014-07-21T00:00:00.000+0000 8 11465 2014-07-28T00:00:00.000+0000 4 106 2014-08-11T00:00:00.000+0000 3 20 2014-08-25T00:00:00.000+0000 176 1 2014-06-30T00:00:00.000+0000 10 5110 2014-04-28T00:00:00.000+0000 2 7847 2014-04-28T00:00:00.000+0000 5 2439 2014-05-12T00:00:00.000+0000 3 5467 2014-05-19T00:00:00.000+0000 10 90 2014-05-26T00:00:00.000+0000 7 717 2014-05-26T00:00:00.000+0000 10 6270 2014-06-02T00:00:00.000+0000 5 102 2014-06-09T00:00:00.000+0000 3 11108 2014-06-16T00:00:00.000+0000 5 3826 2014-06-30T00:00:00.000+0000 7 31 2014-07-14T00:00:00.000+0000 3 5934 2014-07-14T00:00:00.000+0000 6 8277 2014-07-21T00:00:00.000+0000 5 7776 2014-07-28T00:00:00.000+0000 25 54 2014-07-21T00:00:00.000+0000 12 4777 2014-07-28T00:00:00.000+0000 9 267 2014-07-28T00:00:00.000+0000 9 11408 2014-08-18T00:00:00.000+0000 6 10906 2014-08-04T00:00:00.000+0000 4 6331 2014-08-25T00:00:00.000+0000 6 12538 2014-08-18T00:00:00.000+0000 26 11 2014-08-25T00:00:00.000+0000 292 3 2014-08-25T00:00:00.000+0000 4 7456 2014-08-25T00:00:00.000+0000 11 25 2014-08-25T00:00:00.000+0000 3 4642 2014-05-05T00:00:00.000+0000 13 9712 2014-05-05T00:00:00.000+0000 5 10291 2014-06-

As shown by the line chart above, the company with ID "1" has the highest number of weekly engagements. There has been a significant drop in engagements from 512 weekly engagements for the week of July 27th to 176 weekly engagements for the week of August 24th. Given that this company was responsible for most of the weekly engagements, it's possible that a decline in engagement from one company could have an impact on the overall engagement on eLinks. The company with ID "2" has also had a significant decline in engagement recently, from 341 weekly engagements for the week of July 20th, to a low of  47 engagements for the week of August 17th. As mentioned previously, possible reasons for the decline in engagement of specific groups could be due to the emergence of a competitor platform.

The data presented in this section supports the idea that there has been a decline in the engagement of the companies and locations that previously had the highest engagement rates. This supports the hypothesis that some people may have started using a different networking site, however further evidence is required to confirm this hypothesis.

#### Hypothesis 4: eLinks Changed Their Engagement Techniques

The final hypothesis is that eLinks changed their engagement techniques, perhaps by reducing the number of weekly digest emails sent to their users. To assess this hypothesis, we will first look at the number of weekly emails sent.

As shown by the graph below, there is a steady increase in the number of weekly digest emails sent per day. The highest number of emails are sent out on Mondays, with lower numbers of emails being sent out on Tuesday, Wednesday, Thursday and Friday.

In [0]:
%sql 
SELECT 
COUNT(CAST(user_id AS int)) AS weekly_emails_sent, 
DATE_TRUNC('day',CAST(occurred_at AS timestamp)) AS day
FROM emails_csv
WHERE action='sent_weekly_digest'
GROUP BY day
ORDER BY day ASC

weekly_emails_sent day 467 2014-05-01T00:00:00.000+0000 441 2014-05-02T00:00:00.000+0000 760 2014-05-05T00:00:00.000+0000 477 2014-05-06T00:00:00.000+0000 435 2014-05-07T00:00:00.000+0000 479 2014-05-08T00:00:00.000+0000 451 2014-05-09T00:00:00.000+0000 766 2014-05-12T00:00:00.000+0000 491 2014-05-13T00:00:00.000+0000 447 2014-05-14T00:00:00.000+0000 494 2014-05-15T00:00:00.000+0000 467 2014-05-16T00:00:00.000+0000 778 2014-05-19T00:00:00.000+0000 508 2014-05-20T00:00:00.000+0000 459 2014-05-21T00:00:00.000+0000 503 2014-05-22T00:00:00.000+0000 485 2014-05-23T00:00:00.000+0000 802 2014-05-26T00:00:00.000+0000 522 2014-05-27T00:00:00.000+0000 471 2014-05-28T00:00:00.000+0000 524 2014-05-29T00:00:00.000+0000 503 2014-05-30T00:00:00.000+0000 822 2014-06-02T00:00:00.000+0000 538 2014-06-03T00:00:00.000+0000 484 2014-06-04T00:00:00.000+0000 543 2014-06-05T00:00:00.000+0000 524 2014-06-06T00:00:00.000+0000 846 2014-06-09T00:00:00.000+0000 554 2014-06-10T00:00:00.000+0000 501 2014-06-11T00:00:00.000+0000 561 2014-06-12T00:00:00.000+0000 541 2014-06-13T00:00:00.000+0000 876 2014-06-16T00:00:00.000+0000 568 2014-06-17T00:00:00.000+0000 521 2014-06-18T00:00:00.000+0000 584 2014-06-19T00:00:00.000+0000 556 2014-06-20T00:00:00.000+0000 905 2014-06-23T00:00:00.000+0000 585 2014-06-24T00:00:00.000+0000 543 2014-06-25T00:00:00.000+0000 599 2014-06-26T00:00:00.000+0000 575 2014-06-27T00:00:00.000+0000 929 2014-06-30T00:00:00.000+0000 602 2014-07-01T00:00:00.000+0000 561 2014-07-02T00:00:00.000+0000 616 2014-07-03T00:00:00.000+0000 594 2014-07-04T00:00:00.000+0000 958 2014-07-07T00:00:00.000+0000 620 2014-07-08T00:00:00.000+0000 577 2014-07-09T00:00:00.000+0000 632 2014-07-10T00:00:00.000+0000 612 2014-07-11T00:00:00.000+0000 987 2014-07-14T00:00:00.000+0000 638 2014-07-15T00:00:00.000+0000 598 2014-07-16T00:00:00.000+0000 644 2014-07-17T00:00:00.000+0000 632 2014-07-18T00:00:00.000+0000 1012 2014-07-21T00:00:00.000+0000 655 2014-07-22T00:00:00.000+0000 621 2014-07-23T00:00:00.000+0000 658 2014-07-24T00:00:00.000+0000 646 2014-07-25T00:00:00.000+0000 1044 2014-07-28T00:00:00.000+0000 674 2014-07-29T00:00:00.000+0000 645 2014-07-30T00:00:00.000+0000 676 2014-07-31T00:00:00.000+0000 667 2014-08-01T00:00:00.000+0000 1068 2014-08-04T00:00:00.000+0000 695 2014-08-05T00:00:00.000+0000 660 2014-08-06T00:00:00.000+0000 692 2014-08-07T00:00:00.000+0000 678 2014-08-08T00:00:00.000+0000 1096 2014-08-11T00:00:00.000+0000 717 2014-08-12T00:00:00.000+0000 678 2014-08-13T00:00:00.000+0000 714 2014-08-14T00:00:00.000+0000 692 2014-08-15T00:00:00.000+0000 1125 2014-08-18T00:00:00.000+0000 733 2014-08-19T00:00:00.000+0000 700 2014-08-20T00:00:00.000+0000 739 2014-08-21T00:00:00.000+0000 715 2014-08-22T00:00:00.000+0000 1153 2014-08-25T00:00:00.000+0000 746 2014-08-26T00:00:00.000+0000 719 2014-08-27T00:00:00.000+0000 758 2014-08-28T00:00:00.000+0000 735 2014-08-29T00:00:00.000+0000

If we consider the same data at the week level instead of the day level, we can see that there is a gradual increase in the number of weekly digest emails sent out each week, as shown by the bar chart below, which makes sense given the increase in users over time.

In [0]:
%sql 
SELECT 
COUNT(CAST(user_id AS int)) AS weekly_digest_emails_sent, 
DATE_TRUNC('week',CAST(occurred_at AS timestamp)) AS week
FROM emails_csv
WHERE action='sent_weekly_digest'
GROUP BY week
ORDER BY week ASC

weekly_digest_emails_sent week 908 2014-04-28T00:00:00.000+0000 2602 2014-05-05T00:00:00.000+0000 2665 2014-05-12T00:00:00.000+0000 2733 2014-05-19T00:00:00.000+0000 2822 2014-05-26T00:00:00.000+0000 2911 2014-06-02T00:00:00.000+0000 3003 2014-06-09T00:00:00.000+0000 3105 2014-06-16T00:00:00.000+0000 3207 2014-06-23T00:00:00.000+0000 3302 2014-06-30T00:00:00.000+0000 3399 2014-07-07T00:00:00.000+0000 3499 2014-07-14T00:00:00.000+0000 3592 2014-07-21T00:00:00.000+0000 3706 2014-07-28T00:00:00.000+0000 3793 2014-08-04T00:00:00.000+0000 3897 2014-08-11T00:00:00.000+0000 4012 2014-08-18T00:00:00.000+0000 4111 2014-08-25T00:00:00.000+0000

Next, we can consider the number of reengagement emails sent out. Reengagement emails are presumably sent out to users who haven't logged on to eLinks for a while in an attempt to re-engage them. As shown by the bar chart below, there number of reengagement emails increased slightly in the last two weeks of the dataset.

In [0]:
%sql 
SELECT 
COUNT(CAST(user_id AS int)) AS reengagement_emails_sent, 
DATE_TRUNC('week',CAST(occurred_at AS timestamp)) AS week
FROM emails_csv
WHERE action='sent_reengagement_email'
GROUP BY week
ORDER BY week ASC

reengagement_emails_sent week 98 2014-04-28T00:00:00.000+0000 164 2014-05-05T00:00:00.000+0000 175 2014-05-12T00:00:00.000+0000 179 2014-05-19T00:00:00.000+0000 179 2014-05-26T00:00:00.000+0000 199 2014-06-02T00:00:00.000+0000 190 2014-06-09T00:00:00.000+0000 234 2014-06-16T00:00:00.000+0000 187 2014-06-23T00:00:00.000+0000 222 2014-06-30T00:00:00.000+0000 214 2014-07-07T00:00:00.000+0000 226 2014-07-14T00:00:00.000+0000 206 2014-07-21T00:00:00.000+0000 230 2014-07-28T00:00:00.000+0000 206 2014-08-04T00:00:00.000+0000 224 2014-08-11T00:00:00.000+0000 257 2014-08-18T00:00:00.000+0000 263 2014-08-25T00:00:00.000+0000

The data explored in this section does not support the hypothesis that the recent decline in engagement is due to a change in engagement tactics as the data shows a slight increase in the number of engagement and re-engagement emails sent out.

### Analysis of Engagement Strategies

To help eLinks determine how effective their weekly digest emails are at getting their users to engage with their website, we can compare the engagement rate of users who clicked through and did not click through their emails from eLinks. In addition to the analysis in the hypothesis section, this analysis of engagement techniques will inform the recommendations for eLinks.

#### Daily Engagement for Users Who Did Not Click The Weekly Digest Email

The chart below shows the distribution of daily engagement data for all users who did not click through the email on a given day. As shown, the median number of engagements is 8 and there is a large spread in the data with a high number of outliers with a high daily engagement.

In [0]:
%sql 

SELECT 
e.ID_day_engagement,
ct.ID_day_email_clickthrough,
nvl(e.engagements,0) AS user_engagements_day,
nvl(ct.clickthroughs,0) AS user_clickthroughs_day

FROM 
(SELECT 
CONCAT(CAST(user_id AS int),'|',DATE_TRUNC('day',CAST(occurred_at AS timestamp))) AS ID_day_engagement,
COUNT(CAST(user_id AS int)) AS engagements
FROM events_csv 
WHERE event_type = 'engagement'
GROUP BY ID_day_engagement) e

FULL OUTER JOIN (SELECT 
CONCAT(CAST(user_id AS int),'|',DATE_TRUNC('day',CAST(occurred_at AS timestamp))) AS ID_day_email_clickthrough,
COUNT(CAST(user_id AS int)) AS clickthroughs
FROM emails_csv
WHERE action = 'email_clickthrough'
GROUP BY ID_day_email_clickthrough) ct

ON e.ID_day_engagement=ct.ID_day_email_clickthrough

WHERE nvl(ct.clickthroughs,0)=0

ID_day_engagement ID_day_email_clickthrough user_engagements_day user_clickthroughs_day 1012|2014-05-04 00:00:00 null 9 0 1012|2014-05-05 00:00:00 null 21 0 1012|2014-06-04 00:00:00 null 10 0 1012|2014-06-09 00:00:00 null 10 0 1012|2014-06-11 00:00:00 null 10 0 1012|2014-06-16 00:00:00 null 13 0 1012|2014-06-19 00:00:00 null 10 0 1012|2014-06-24 00:00:00 null 5 0 1012|2014-08-08 00:00:00 null 7 0 1012|2014-08-09 00:00:00 null 12 0 1012|2014-08-12 00:00:00 null 11 0 1012|2014-08-13 00:00:00 null 3 0 1012|2014-08-17 00:00:00 null 14 0 1012|2014-08-23 00:00:00 null 15 0 1014|2014-08-01 00:00:00 null 9 0 1018|2014-07-18 00:00:00 null 23 0 1019|2014-06-24 00:00:00 null 4 0 1019|2014-07-03 00:00:00 null 27 0 1019|2014-07-04 00:00:00 null 3 0 1019|2014-07-09 00:00:00 null 6 0 1019|2014-07-15 00:00:00 null 6 0 1019|2014-07-16 00:00:00 null 6 0 1019|2014-07-26 00:00:00 null 6 0 1019|2014-08-03 00:00:00 null 3 0 1019|2014-08-07 00:00:00 null 4 0 1019|2014-08-27 00:00:00 null 8 0 101|2014-05-16 00:00:00 null 20 0 101|2014-05-23 00:00:00 null 4 0 101|2014-05-31 00:00:00 null 14 0 101|2014-06-06 00:00:00 null 8 0 101|2014-06-26 00:00:00 null 11 0 101|2014-06-27 00:00:00 null 14 0 101|2014-08-01 00:00:00 null 5 0 101|2014-08-06 00:00:00 null 9 0 101|2014-08-09 00:00:00 null 8 0 101|2014-08-13 00:00:00 null 12 0 101|2014-08-20 00:00:00 null 40 0 101|2014-08-22 00:00:00 null 23 0 101|2014-08-23 00:00:00 null 41 0 101|2014-08-26 00:00:00 null 7 0 101|2014-08-27 00:00:00 null 11 0 101|2014-08-28 00:00:00 null 13 0 101|2014-08-30 00:00:00 null 11 0 1024|2014-06-13 00:00:00 null 8 0 1024|2014-06-20 00:00:00 null 11 0 1024|2014-06-23 00:00:00 null 5 0 1024|2014-06-28 00:00:00 null 8 0 1033|2014-06-04 00:00:00 null 8 0 1033|2014-06-07 00:00:00 null 8 0 1033|2014-07-16 00:00:00 null 16 0 1035|2014-07-03 00:00:00 null 6 0 1047|2014-05-09 00:00:00 null 10 0 1047|2014-05-14 00:00:00 null 15 0 1047|2014-05-20 00:00:00 null 5 0 1047|2014-05-23 00:00:00 null 8 0 1047|2014-05-26 00:00:00 null 13 0 1047|2014-05-27 00:00:00 null 5 0 1047|2014-05-28 00:00:00 null 8 0 1047|2014-05-29 00:00:00 null 8 0 1047|2014-05-30 00:00:00 null 8 0 1047|2014-06-06 00:00:00 null 14 0 1047|2014-06-10 00:00:00 null 11 0 1047|2014-07-22 00:00:00 null 15 0 1047|2014-07-29 00:00:00 null 14 0 1047|2014-07-30 00:00:00 null 33 0 1047|2014-08-01 00:00:00 null 13 0 1047|2014-08-05 00:00:00 null 8 0 1047|2014-08-08 00:00:00 null 16 0 1047|2014-08-14 00:00:00 null 29 0 1047|2014-08-18 00:00:00 null 15 0 1047|2014-08-20 00:00:00 null 5 0 1047|2014-08-21 00:00:00 null 11 0 1047|2014-08-22 00:00:00 null 12 0 1047|2014-08-26 00:00:00 null 20 0 1047|2014-08-27 00:00:00 null 9 0 1047|2014-08-29 00:00:00 null 8 0 10522|2014-05-02 00:00:00 null 6 0 1055|2014-05-16 00:00:00 null 4 0 1055|2014-07-04 00:00:00 null 6 0 1055|2014-07-09 00:00:00 null 2 0 1055|2014-07-15 00:00:00 null 10 0 1055|2014-07-16 00:00:00 null 7 0 1055|2014-07-17 00:00:00 null 4 0 1055|2014-07-23 00:00:00 null 10 0 1055|2014-07-25 00:00:00 null 3 0 1055|2014-07-30 00:00:00 null 4 0 10612|2014-05-01 00:00:00 null 12 0 1061|2014-06-23 00:00:00 null 5 0 1061|2014-06-27 00:00:00 null 4 0 1061|2014-06-29 00:00:00 null 4 0 1061|2014-07-04 00:00:00 null 5 0 1061|2014-07-06 00:00:00 null 6 0 1061|2014-07-07 00:00:00 null 4 0 1061|2014-07-09 00:00:00 null 7 0 1061|2014-07-12 00:00:00 null 11 0 1061|2014-07-14 00:00:00 null 3 0 1061|2014-07-19 00:00:00 null 4 0 1061|2014-07-29 00:00:00 null 11 0 1061|2014-08-01 00:00:00 null 13 0 1061|2014-08-02 00:00:00 null 5 0 1061|2014-08-06 00:00:00 null 10 0 1063|2014-05-21 00:00:00 null 3 0 1063|2014-05-28 00:00:00 null 11 0 1063|2014-06-25 00:00:00 null 9 0 1063|2014-06-27 00:00:00 null 8 0 1063|2014-06-28 00:00:00 null 11 0 1063|2014-06-30 00:00:00 null 9 0 1063|2014-07-04 00:00:00 null 10 0 1063|2014-07-10 00:00:00 null 12 0 1063|2014-07-11 00:00:00 null 7 0 1063|2014-07-30 00:00:00 null 10 0 10736|2014-05-09 00:00:00 null 3 0 1078|2014-05-22 00:00:00 null 10 0 1078|2014-05-31 00:00:00 null 5

#### Daily Engagement for Users Who Did Click Through the Weekly Digest Email

Next, we can consider the daily engagement of all users who did click through the email they received on a given day. Interestingly, the daily engagement of those users who did click through the weekly digest email is lower than the distribution of the daily engagement of users who did not click through the weekly digest email, with a median of 3 daily engagements. The upper and lower quartile bounds of the distribution of the daily engagement data for users who did click on the email is much smaller than the data for users who did not click through the email. Similar to the distribution of the data for users who did not click through the email, there are a large number of outliers with a high daily engagement rate, with the highest daily engagement rate being 65 engagement activities for a given day.

In [0]:
%sql 

SELECT 
e.ID_day_engagement,
ct.ID_day_email_clickthrough,
nvl(e.engagements,0) AS user_engagements_day,
nvl(ct.clickthroughs,0) AS user_clickthroughs_day

FROM 
(SELECT 
CONCAT(CAST(user_id AS int),'|',DATE_TRUNC('day',CAST(occurred_at AS timestamp))) AS ID_day_engagement,
COUNT(CAST(user_id AS int)) AS engagements
FROM events_csv 
WHERE event_type = 'engagement'
GROUP BY ID_day_engagement) e

FULL OUTER JOIN (SELECT 
CONCAT(CAST(user_id AS int),'|',DATE_TRUNC('day',CAST(occurred_at AS timestamp))) AS ID_day_email_clickthrough,
COUNT(CAST(user_id AS int)) AS clickthroughs
FROM emails_csv
WHERE action = 'email_clickthrough'
GROUP BY ID_day_email_clickthrough) ct

ON e.ID_day_engagement=ct.ID_day_email_clickthrough

WHERE nvl(ct.clickthroughs,0)>=1

ID_day_engagement ID_day_email_clickthrough user_engagements_day user_clickthroughs_day 655|2014-07-31 00:00:00 655|2014-07-31 00:00:00 8 1 2050|2014-07-25 00:00:00 2050|2014-07-25 00:00:00 29 1 4032|2014-06-23 00:00:00 4032|2014-06-23 00:00:00 12 1 4163|2014-05-05 00:00:00 4163|2014-05-05 00:00:00 5 1 5508|2014-07-09 00:00:00 5508|2014-07-09 00:00:00 13 1 5597|2014-08-15 00:00:00 5597|2014-08-15 00:00:00 4 1 5633|2014-06-02 00:00:00 5633|2014-06-02 00:00:00 29 1 6074|2014-06-10 00:00:00 6074|2014-06-10 00:00:00 11 1 null 6401|2014-07-04 00:00:00 0 1 null 7089|2014-05-16 00:00:00 0 1 null 7455|2014-05-14 00:00:00 0 1 null 7488|2014-07-10 00:00:00 0 1 null 7846|2014-05-12 00:00:00 0 1 null 8209|2014-06-04 00:00:00 0 1 null 8259|2014-06-05 00:00:00 0 1 null 8378|2014-05-19 00:00:00 0 1 null 8614|2014-06-23 00:00:00 0 1 null 9353|2014-06-06 00:00:00 0 1 null 9781|2014-06-24 00:00:00 0 1 null 9993|2014-08-11 00:00:00 0 1 null 11030|2014-07-30 00:00:00 0 1 null 11414|2014-07-09 00:00:00 0 1 null 12955|2014-07-11 00:00:00 0 1 null 12976|2014-07-04 00:00:00 0 1 null 13387|2014-07-14 00:00:00 0 1 null 13406|2014-06-30 00:00:00 0 1 null 14042|2014-07-24 00:00:00 0 1 null 14570|2014-08-22 00:00:00 0 1 null 12313|2014-05-16 00:00:00 0 1 null 12522|2014-05-19 00:00:00 0 1 null 12924|2014-05-27 00:00:00 0 1 null 13093|2014-05-31 00:00:00 0 1 null 13581|2014-06-08 00:00:00 0 1 null 13871|2014-06-14 00:00:00 0 1 null 14295|2014-06-21 00:00:00 0 1 null 14351|2014-06-22 00:00:00 0 1 null 14478|2014-06-23 00:00:00 0 1 null 14518|2014-06-24 00:00:00 0 1 373|2014-07-08 00:00:00 373|2014-07-08 00:00:00 9 1 1297|2014-07-25 00:00:00 1297|2014-07-25 00:00:00 16 1 3584|2014-07-30 00:00:00 3584|2014-07-30 00:00:00 9 1 4092|2014-06-11 00:00:00 4092|2014-06-11 00:00:00 6 1 4580|2014-08-19 00:00:00 4580|2014-08-19 00:00:00 5 1 5074|2014-07-21 00:00:00 5074|2014-07-21 00:00:00 15 1 5120|2014-05-27 00:00:00 5120|2014-05-27 00:00:00 8 1 5249|2014-05-05 00:00:00 5249|2014-05-05 00:00:00 5 1 5462|2014-06-16 00:00:00 5462|2014-06-16 00:00:00 4 1 5633|2014-06-30 00:00:00 5633|2014-06-30 00:00:00 33 1 5706|2014-08-20 00:00:00 5706|2014-08-20 00:00:00 5 1 null 7018|2014-07-03 00:00:00 0 1 null 7673|2014-07-22 00:00:00 0 1 null 7941|2014-07-16 00:00:00 0 1 null 8514|2014-06-05 00:00:00 0 1 null 9993|2014-08-18 00:00:00 0 1 null 10403|2014-08-20 00:00:00 0 1 null 11293|2014-08-19 00:00:00 0 1 null 13248|2014-08-28 00:00:00 0 1 null 13631|2014-06-19 00:00:00 0 1 null 13974|2014-06-25 00:00:00 0 1 null 14441|2014-07-03 00:00:00 0 1 null 16325|2014-07-28 00:00:00 0 1 null 11854|2014-05-06 00:00:00 0 1 null 12700|2014-05-23 00:00:00 0 1 null 13662|2014-06-09 00:00:00 0 1 4|2014-05-13 00:00:00 4|2014-05-13 00:00:00 4 1 466|2014-05-27 00:00:00 466|2014-05-27 00:00:00 24 1 527|2014-06-30 00:00:00 527|2014-06-30 00:00:00 14 1 1374|2014-07-03 00:00:00 1374|2014-07-03 00:00:00 4 1 3119|2014-07-08 00:00:00 3119|2014-07-08 00:00:00 2 1 3484|2014-06-05 00:00:00 3484|2014-06-05 00:00:00 12 1 3821|2014-05-23 00:00:00 3821|2014-05-23 00:00:00 17 1 4282|2014-05-08 00:00:00 4282|2014-05-08 00:00:00 7 1 4513|2014-08-11 00:00:00 4513|2014-08-11 00:00:00 2 1 5291|2014-07-29 00:00:00 5291|2014-07-29 00:00:00 5 1 5548|2014-06-05 00:00:00 5548|2014-06-05 00:00:00 9 1 6018|2014-06-02 00:00:00 6018|2014-06-02 00:00:00 2 1 null 7056|2014-05-15 00:00:00 0 1 null 7484|2014-07-02 00:00:00 0 1 null 7549|2014-07-11 00:00:00 0 1 null 8216|2014-07-09 00:00:00 0 1 null 8453|2014-07-16 00:00:00 0 1 null 8469|2014-07-30 00:00:00 0 1 null 8660|2014-07-22 00:00:00 0 1 null 9160|2014-07-07 00:00:00 0 1 null 9488|2014-08-12 00:00:00 0 1 null 11014|2014-06-24 00:00:00 0 1 null 11323|2014-07-08 00:00:00 0 1 null 14885|2014-08-14 00:00:00 0 1 null 15283|2014-08-28 00:00:00 0 1 null 16949|2014-08-27 00:00:00 0 1 null 11887|2014-05-06 00:00:00 0 1 null 12077|2014-05-11 00:00:00 0 1 null 12212|2014-05-13 00:00:00 0 1 null 12227|2014-05-13 00:00:00 0 1 null 13621|2014-06-09 00:00:00 0 1 null 13658|2014-06-09 00:0

The data presented above shows an unexpected trend, that on a given day users who clicked through the weekly digest email that day tend to have a lower engagement on that day compared with the data for users that did not click through the weekly digest email. To examine the effectiveness of the weekly digest email at a high level, we can also consider the data at the user level to see if there is any correlation between the number of times a user has clicked through the weekly digest email and the number of times they have engaged on the platform.

The scatterplot below shows the relationship shows the relationship between the total number of clickthroughs and total number of engagements actions at the user level. As shown, there does not appear to be a strong relationship between the total clickthroughs and total engagement actions per user, which suggests that sending the weekly digest emails may not be the most effective engagement strategy that eLinks could use.

In [0]:
%sql 

SELECT 
e.user_id, 
ct.user_id,
NVL(ct.total_clickthroughs,0) AS total_clickthroughs,
NVL(e.total_engagements,0) AS total_engagements

FROM (SELECT 
CAST(user_id AS int),
COUNT(CAST(user_id AS int)) AS total_engagements
FROM events_csv 
WHERE event_type = 'engagement'
GROUP BY user_id) e 

FULL OUTER JOIN 

(SELECT 
CAST(user_id AS int), 
COUNT(CAST(user_id AS int)) AS total_clickthroughs
FROM emails_csv
WHERE action = 'email_clickthrough'
GROUP BY user_id) ct

ON e.user_id=ct.user_id

user_id user_id total_clickthroughs total_engagements 4 4 4 93 8 8 1 36 11 11 2 126 17 17 1 55 19 19 1 71 20 20 3 68 22 22 3 357 30 30 1 67 49 49 1 5 59 59 3 257 64 64 2 47 66 null 0 43 78 78 4 75 80 80 2 73 83 83 1 90 86 86 3 118 98 98 2 44 101 101 6 401 124 124 2 17 128 128 4 124 134 134 4 216 136 136 2 222 140 140 1 70 150 150 4 180 163 163 3 276 170 170 1 28 171 171 2 12 172 172 2 118 175 175 4 92 181 181 2 13 182 182 1 89 201 201 3 51 203 203 4 363 209 null 0 22 214 214 1 62 215 215 3 162 226 226 2 16 227 null 0 50 232 232 3 204 239 239 3 128 244 244 1 9 247 247 2 100 253 253 1 136 259 259 2 9 260 260 3 228 262 262 1 64 264 264 3 76 270 270 1 109 271 271 1 7 285 285 3 234 286 286 1 23 287 287 2 299 300 300 2 30 301 301 2 35 306 306 1 90 319 319 5 150 323 323 2 22 332 332 2 10 335 335 1 3 341 341 1 5 345 345 2 40 346 346 2 32 347 347 4 406 349 349 1 18 353 353 3 42 359 359 2 150 363 363 1 92 365 365 1 124 366 366 4 417 372 372 3 343 373 373 5 52 376 376 2 169 381 381 4 113 383 383 1 76 388 388 4 71 390 390 1 110 402 402 2 56 403 403 1 76 404 404 2 12 406 406 1 27 428 428 3 152 429 429 5 141 430 430 4 23 434 434 1 78 437 437 3 69 441 441 1 77 451 451 1 82 457 457 2 82 460 460 1 13 461 461 2 109 462 462 2 143 466 466 2 90 471 471 1 4 479 479 1 19 489 489 1 54 491 491 1 8 492 492 4 189 493 493 3 37 498 498 1 13 501 501 3 262 503 503 2 28 521 521 3 334 523 523 2 304 527 527 2 65 529 529 3 254 535 535 2 16 546 546 1 7 551 551 2 200 552 552 3 211 555 555 4 197 563 563 1 72 565 565 1 76 566 566 1 57 575 575 1 112 576 576 1 199 581 581 2 82 586 586 2 104 589 589 3 199 591 591 4 307 597 597 3 272 602 602 4 53 614 614 1 37 621 621 4 132 633 633 2 23 643 643 1 159 644 644 2 94 655 655 1 8 660 660 4 79 668 668 3 67 673 673 2 35 684 684 2 120 688 688 1 89 689 689 3 24 698 698 2 24 700 700 2 110 702 702 3 202 703 703 3 21 707 707 2 55 709 709 1 18 714 714 1 137 719 719 3 236 720 720 2 34 726 726 1 6 729 729 2 130 745 745 3 289 746 746 3 63 753 753 1 79 756 756 1 45 759 759 3 220 761 761 4 317 771 771 2 152 772 772 2 18 773 773 1 10 775 775 3 79 780 780 3 87 783 783 1 62 792 792 2 21 800 800 1 22 802 802 2 22 806 806 2 16 819 819 4 131 822 822 6 131 823 823 1 20 825 825 4 240 827 827 1 86 831 831 1 30 836 836 3 45 857 857 2 292 864 864 1 27 899 899 1 59 908 908 4 27 914 914 2 167 915 915 3 31 916 916 1 123 919 919 3 197 927 927 2 110 931 931 1 171 934 934 2 28 935 935 2 15 937 937 2 35 949 949 2 51 952 952 3 75 959 959 4 46 960 960 1 11 966 966 1 78 973 973 3 221 984 984 2 27 1012 1012 3 188 1014 1014 2 31 1018 1018 2 49 1019 1019 2 93 1024 1024 1 43 1033 1033 2 54 1035 1035 1 14 1047 1047 4 415 1055 1055 4 73 1061 1061 2 107 1063 1063 3 145 1078 1078 2 81 1080 1080 4 243 1085 1085 1 9 1090 1090 2 16 1104 1104 3 21 1108 1108 3 89 1114 1114 1 97 1119 1119 1 3 1121 1121 2 137 1126 1126 1 48 1127 1127 3 23 1128 1128 2 38 1135 1135 4 85 1139 1139 2 112 1147 1147 3 262 1154 1154 2 247 1164 1164 1 4 1165 1165 2 65 1170 1170 2 178 1179 1179 1 14 1194 1194 1 67 1208 1208 5 56 1215 1215 3 253 1218 1218 3 46 1220 1220 2 84 1225 1225 2 86 1238 1238 1 4 1252 1252 1 40 1260 null 0 6 1271 1271 1 41 1275 1275 2 273 1283 1283 3 135 1287 1287 1 150 1293 1293 5 54 1295 1295 2 39 1297 1297 2 215 1300 1300 2 129 1301 1301 1 133 1306 1306 3 111 1308 1308 1 91 1311 1311 4 372 1312 1312 2 199 1319 1319 3 212 1322 1322 3 14 1326 1326 6 111 1345 1345 1 35 1346 1346 2 43 1351 1351 3 391 1352 1352 2 14 1361 1361 1 43 1365 1365 2 8 1367 null 0 20 1371 1371 3 162 1373 1373 3 206 1374 1374 2 30 1377 1377 1 4 1380 1380 3 291 1381 1381 3 268 1384 1384 4 304 1386 1386 2 12 1388 1388 2 162 1400 1400 1 10 1401 1401 3 56 1403 1403 2 45 1410 1410 3 145 1412 1412 2 247 1415 1415 3 217 1419 1419 2 34 1422 1422 1 8 1423 1423 3 161 1424 1424 4 311 1425 1425 3 41 1427 1427 3 89 1428 null 0 119 1439 1439 3 214 1440 1440 1 52 1445 1445 2 21 1446 1446 1 2 1450 1450 2 25 1461 1461 2 27 1463 1463 1 2 1465 1465 3 41 1468 1468 1 22 1477 1477 2 24 1481 1481 1 2 1482 1482 3 67 1488 1488 

### Summary of Findings

#### Hypothesis 1: Users Who Joined eLinks a Long Time Ago Are No Longer Engaging
To test the hypothesis that older users are no longer engaging on eLinks, the relationship between the number of days on the platform and days since last engagement was assessed. Although a few clusters of users could be identified, this assessment showed that there was no linear relationship between time on platform and engagement. Based on the evidence collected, the hypothesis that users who joined eLinks a long time ago are no longer engaging can be rejected.
#### Hypothesis 2: Certain Elements of the Website Were Non-Functional, Resulting in a Decline in Engagement
To test the hypothesis that certain elements of the website weren't working, preventing users from engaging, the number of engagement activities was assessed at the weekly scale. The data showed a drop in all engagement activities in August, with the largest drop being the number of people loading the home screen. To ensure that the website was functioning normally during August, the number of daily sign up activities was also assessed. As the sign up activity looked normal during the end of August it was concluded that the website was fully functional and people were not engaging on eLinks for another reason. Based on the evidence collected, the hypothesis that certain elements of the website were non-function can be rejected.
#### Hypothesis 3: Some Users Have Switched Over to a Different Platform
Without data on user activity outside of their use of eLinks, the hypothesis that some users had switched over to a different platform was assessed by looking for patterns in the groups of users that were no longer engaging on eLinks. The engagement was plotted over time for all locations and Germany and France were found to have the largest drop in engagement. The engagement data was also examined by company, and company 1 and 2 was found to have the largest decline in engagement. This evidence suggests support for the hypothesis that users have switched to a different platform, however data on users activity on other platforms is required to confidently conclude that those users have switched to a different platform.
#### Hypothesis 4: eLinks Changed Their Engagement Techniques
To asses whether eLinks had changed their engagement techniques, the number of weekly digest emails and re-engagement emails sent were examined at the daily and weekly scale. This analysis showed a slight increase in the number of these emails sent, but no drastic changes that would explain a drop in engagement. Therefore, this hypothesis can be rejected.

In summary, there is the most evidence to support hypothesis that some users have switched over to a different platform. Without more information on the activity of users outside of their use of eLinks, it's difficult to confidently conclude that users have switched to a different platform. It is therefore recommended to conduct market analysis to assess whether the decline in engagement in Germany, France and Company and company 1 and 2 is due to a switch to a competitor.

### Recommendations

#### Recommendation 1: Conduct a Follow-up Study
As mentioned above, based on the evidence examined in this analysis it is likely that users from company 1 and 2 as well as users from Germany and France may have switched to a different networking platform. Given the non-direct evidence, it's not possible to accept this hypothesis with confidence. A follow-up study should be conducted to investigate the drop in users at companies 1 and 2 as well as users in Germany and France. This study should look for evidence that would indicate why those users have stopped engaging on eLinks and should gather additional data beyond what was explored in this analysis. The follow-up study could address the following areas: 
* Market Research
* Engagement Techniques

#### Recommendation 2: Conduct Market Research on Competitors
As part of the follow-up study, market research should be conducted to gather information on competitor networking platforms. Specific questions that the research to look to answer are: 
1. What networking platforms are leaders in the industries of companies 1 and 2? 
2. What features do those platforms have that eLinks does not currently have? 
3. What networking platforms are leaders or emerging platforms in Western Europe (Germany and France specifically)?
4. If possible, gather online survey data directly from non-engaged users to gain insight into whether they are using a different networking platform and what features they like from that platform
5. What techniques do competitors use to encourage users to engage on their platforms?

#### Recommendation 3: Utilize Effective Engagement Techniques
The Analysis of Engagement Strategies section of this analysis showed that the weekly digest emails are not an effective engagement strategy, as users who clicked through those emails did not have higher engagement rates. To ensure that any engagement strategy used is effective at getting users to interact on the eLinks, it is recommended that eLinks conducts research into different engagement techniques so they can utilize the most effective techniques. This research should aim to answer the following questions: 
1. What engagement technique leads to the highest engagement? Is it a weekly digest email? Is it biweekly highlights of the popular posts in their feed? 
2. What is the best time to send emails to users? 
3. What is the best weekday to send emails to users?

### Conclusion

In conclusion, eLinks has experienced a fairly significant drop in engagement that started at the end of July. Based on this initial analysis, there is evidence to suggest that some users have switched to a different networking platform. It is therefore recommended that eLinks conducts a follow-up study to conduct market research into competitor networking sites, gather additional information from their un-engaged users and research effective engagement techniques.